Support GlobalAccess for gce Internal Loadbalancers

prameshj opened the pull request:
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR adds Support for GlobalAccess for gce Internal Loadbalancers.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Added support to specify a global-access annotation for gce ILB.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

prameshj committed:
Added support for ILB Global Access

Uses beta api when global access is enabled and ga api otherwise
Deprecate the older load-balancer-type annotation

Unit test to verify global access
Also added tests to verify new and old load-balancer-type annotations

Addressed review comments, staticcheck fixes

fixed typo

mattjmcnaughton reviewed (commented):
Thanks for your pr :)

Is there an issue or KEP discussing adding support for `GlobalAccess` for gce internal load balancers? This seems like a fairly large, user-facing change, and to be honest, I don't have a ton of context on what global access is or the ramifications of adding it.

prameshj commented:
> Thanks for your pr :)
> 
> Is there an issue or KEP discussing adding support for `GlobalAccess` for gce internal load balancers? This seems like a fairly large, user-facing change, and to be honest, I don't have a ton of context on what global access is or the ramifications of adding it.

Thanks Matt. I just opened issue - https://github.com/kubernetes/kubernetes/issues/81555 and added some context.

GlobalAccess is an option GCE k8s users can pass in by adding the appropriate annotation. The code changes for this are in commit:
https://github.com/kubernetes/kubernetes/pull/81549/commits/5c1dde77bb7f2421c70251d294695b63a3b3784a

The rest are all autogenerated vendor/go.mod changes.

mattjmcnaughton commented:
> > Thanks for your pr :)
> > Is there an issue or KEP discussing adding support for `GlobalAccess` for gce internal load balancers? This seems like a fairly large, user-facing change, and to be honest, I don't have a ton of context on what global access is or the ramifications of adding it.
> 
> Thanks Matt. I just opened issue - #81555 and added some context.
> 
> GlobalAccess is an option GCE k8s users can pass in by adding the appropriate annotation. The code changes for this are in commit:
> [5c1dde7](https://github.com/kubernetes/kubernetes/commit/5c1dde77bb7f2421c70251d294695b63a3b3784a)
> 
> The rest are all autogenerated vendor/go.mod changes.

I wonder, would it be possible to make the vendor changes in a separate pr? I think it could make it a lot easier for reviewers to focus on the application logic changes.

cblecker commented:
@mattjmcnaughton Separate commit, sure. Not a separate PR. Vendor changes that are related to a feature should be in the same PR, so if it's reverted, all of it gets reverted.

prameshj commented:
> Matt McNaughton Separate commit, sure. Not a separate PR. Vendor changes that are related to a feature should be in the same PR, so if it's reverted, all of it gets reverted.

Yes, the changes are in a separate commit currently. I put the vendor changes in the same PR so that the build/tests will succeed too.

mattjmcnaughton commented:
On Sun, Aug 18, 2019 at 01:07:29PM -0700, Christoph Blecker wrote:
> @mattjmcnaughton Separate commit, sure. Not a separate PR. Vendor changes that are related to a feature should be in the same PR, so if it's reverted, all of it gets reverted.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/kubernetes/kubernetes/pull/81549#issuecomment-522351115

Ah yes, good point :)

prameshj commented:
/test pull-kubernetes-e2e-gce

prameshj commented:
@freehan @bowei

prameshj commented:
/assign @bowei

prameshj commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -745,3 +746,121 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+
+func (g *Cloud) ensureInternalForwardingRule(existingFwdRule *compute.ForwardingRule, expectedRuleIntf *fwdRuleIntf, options *ILBOptions) (err error) {
+	expectedRuleIntf.constructRules(options)
+	lbName := expectedRuleIntf.name
```
looks like this is just for logging, I would prefer just using `expectedRuleIntf.name`

prameshj replied:
done

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -745,3 +746,121 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+		needsDelete := existingAPIVersion != newAPIVersion
+		if !needsDelete {
+			// api version matches, check if rule parameters are different
+			needsDelete, err = g.fwdRuleCompare(existingFwdRule, expectedRuleIntf, options, newAPIVersion)
```
not sure about this logic, seems like we should just update?

bowei replied:
also, don't need to delete /create just to go between api versions

prameshj replied:
There is no update api for forwardingRule. So far, since globalAccess is the only beta feature, if the created rule has apiversion beta or if new rule needs beta and the 2 versions aren't the same, it implies deletion. I simplified this to explicitly check if the 2 rules are equal and delete otherwise.

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -745,3 +746,121 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+	}
+	existingBetaRule, err := g.GetBetaRegionForwardingRule(existingRule.Name, g.region)
+	if err != nil {
+		// cannot fetch rule, assume the rules are different
```
We do need to return a bool along with the error. If error is non-nil, the caller will not use the bool return value.

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -745,3 +746,121 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+}
+
+func (f *fwdRuleIntf) constructRules(options *ILBOptions) {
+	f.gaRule = &compute.ForwardingRule{
```
done

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -745,3 +746,121 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+		IPProtocol:          f.ipProtocol,
+		LoadBalancingScheme: f.lbScheme,
+	}
+	if f.subnetwork != "" {
```
We set the Network field only if subnetwork is empty.

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -745,3 +746,121 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+	return &ILBOptions{EnableGlobalAccess: globalAccess}
+}
+
+type fwdRuleIntf struct {
```
done

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -745,3 +746,121 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
 }
+
+func getILBOptions(svc *v1.Service) *ILBOptions {
+	globalAccess := GetLoadBalancerAnnotationGlobalAccess(svc)
```
I had added this so it is easier to check for more options when we add them.. fixed.

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -700,14 +693,22 @@ func backendSvcEqual(a, b *compute.BackendService) bool {
+		a.IPProtocol == b.IPProtocol &&
+		a.LoadBalancingScheme == b.LoadBalancingScheme &&
+		equalStringSets(a.Ports, b.Ports) &&
+		a.BackendService == b.BackendService && (a.AllowGlobalAccess == b.AllowGlobalAccess)
```
This field is not visible in the GA api.

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -115,46 +125,29 @@ func (g *Cloud) ensureInternalLoadBalancer(clusterName, clusterID string, svc *v
+		subnetwork: subnetworkURL,
+		network:    g.networkURL,
+	}
+	err = g.ensureInternalForwardingRule(existingFwdRule, expectedfwdRuleIntf, options)
```
done

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_annotations.go:
```diff
@@ -116,3 +129,19 @@ func GetServiceNetworkTier(service *v1.Service) (cloud.NetworkTier, error) {
+// GetLoadBalancerAnnotationGlobalAccess returns if global access is enabled
+// for the given loadbalancer service.
+func GetLoadBalancerAnnotationGlobalAccess(service *v1.Service) bool {
+	if val, exists := service.Annotations[ServiceAnnotationILBGlobalAccess]; exists && val == "true" {
```
done

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_annotations.go:
```diff
@@ -70,7 +78,12 @@ func GetLoadBalancerAnnotationType(service *v1.Service) (LoadBalancerType, bool)
 	v = LoadBalancerType(l)
 	if !ok {
-		return v, false
+		// Check for deprecated annotation key
```
done

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_annotations.go:
```diff
@@ -22,7 +22,7 @@ import (
 
 	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
done

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -18,13 +18,15 @@ package gce
+	computebeta "google.golang.org/api/compute/v0.beta"
 	compute "google.golang.org/api/compute/v1"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
done

prameshj commented:
> @mattjmcnaughton Separate commit, sure. Not a separate PR. Vendor changes that are related to a feature should be in the same PR, so if it's reverted, all of it gets reverted.

I opened https://github.com/kubernetes/kubernetes/pull/81871 for the vendor changes since it is not tied to this one PR. We have a few followups that will need it.

prameshj commented:
/assign @bowei

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -747,3 +734,203 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+}
+
+func (f *forwardingRuleComposite) Equal(other *forwardingRuleComposite) bool {
+	// IPAddress being an output-only field, equality check ignored if one is empty.
```
if it is output only, shouldn't we just omit from equality?

prameshj replied:
Actually the comment is not correct. We do hold the ip address based on the subnet/user-input. So user could specify a different IP and then it is an update.  Removed the comment.

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -747,3 +734,203 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+		}
+		return frc, err
+	default:
+		return nil, fmt.Errorf("Invalid object type %v to compute ForwardingRuleComposite from", v)
```
type is %T

prameshj replied:
done

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -747,3 +734,203 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+// ToForwardingRuleComposite converts a compute beta or GA ForwardingRule into the composite type
+func ToForwardingRuleComposite(rule interface{}) (frc *forwardingRuleComposite, err error) {
+	switch v := rule.(type) {
+	case *compute.ForwardingRule:
```
use a type switch https://tour.golang.org/methods/16:
```
switch fr := rule.(type) {
  case *compute.ForwardingRule:
    ...
  case *computebeta.ForwardingRule:
    ...
```

prameshj replied:
This is already using type switch. Removed unnecessary typecast.

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -747,3 +734,203 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+
+type forwardingRuleDescription struct {
+	ServiceName string       `json:"kubernetes.io/service-name"`
+	APIVersion  meta.Version `json:"apiVersion,omitempty"`
```
make this consistent
`json:"kubernetes.io/apiVersion,omitempty"`

prameshj replied:
done

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -747,3 +734,203 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+		if err != nil {
+			return fmt.Errorf("Failed to fetch APIVersion from forwardingRule %s, err %v", existingFwdRule.Name, err)
+		}
+		var oldFrc *forwardingRuleComposite
```
oldFRC

prameshj replied:
done

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -747,3 +734,203 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+			oldFrc, err = ToForwardingRuleComposite(existingFwdRule)
+		}
+		if err != nil {
+			return fmt.Errorf("Failed to fetch ForwardingRuleComposite from rule %s, err: %v", existingFwdRule.Name, err)
```
don't wrap the error

prameshj replied:
done

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -747,3 +734,203 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+		if err != nil {
+			return fmt.Errorf("Failed to fetch ForwardingRuleComposite from rule %s, err: %v", existingFwdRule.Name, err)
+		}
+		if oldFrc.Equal(newFrc) {
```
klog.V(4).Infof("oldFRC == newFRC, no updates needed (oldFRC == %+v)", oldFRC)

prameshj replied:
done

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -747,3 +734,203 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+		if oldFrc.Equal(newFrc) {
+			return nil
+		}
+		// delete forwarding rule
```
remove this comment

prameshj replied:
done

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -747,3 +734,203 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+		err = g.CreateRegionForwardingRule(gaRule, g.region)
+	}
+	klog.V(2).Infof("ensureInternalLoadBalancer(%v): created forwarding rule, err : %s", newFrc.name, err)
+	return nil
```
doesn't this have to be `return err`

prameshj replied:
good catch, fixed.

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal_test.go:
```diff
@@ -917,3 +951,200 @@ func TestEnsureInternalLoadBalancerDeletedSubsetting(t *testing.T) {
+		AllowGlobalAccess:   true,
+	}
+	err = gce.CreateRegionForwardingRule(gaRule, gce.region)
+	assert.NoError(t, err)
```
don't use aserrtion, use

```
if err != nil {
  t.Errorf(...)
}
```

prameshj replied:
done

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer.go:
```diff
@@ -152,7 +152,11 @@ func (g *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, svc
+		klog.Errorf("Failed to EnsureLoadBalancer(%v, %v, %v, %v, %v), err: %v", clusterName, svc.Namespace, svc.Name, loadBalancerName, g.region, err)
+		return status, err
+	}
+	klog.V(4).Infof("EnsureLoadBalancer(%v, %v, %v, %v, %v): done ensuring loadbalancer.", clusterName, svc.Namespace, svc.Name, loadBalancerName, g.region)
```
done

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer.go:
```diff
@@ -152,7 +152,11 @@ func (g *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, svc
 	}
-	klog.V(4).Infof("EnsureLoadBalancer(%v, %v, %v, %v, %v): done ensuring loadbalancer. err: %v", clusterName, svc.Namespace, svc.Name, loadBalancerName, g.region, err)
+	if err != nil {
+		klog.Errorf("Failed to EnsureLoadBalancer(%v, %v, %v, %v, %v), err: %v", clusterName, svc.Namespace, svc.Name, loadBalancerName, g.region, err)
```
done

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -747,3 +734,205 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
+func (f *forwardingRuleComposite) ToGA() (*compute.ForwardingRule, error) {
+	descStr, err := f.description.marshal()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to compute description for beta forwarding rule %s, err: %v", f.name, err)
```
beta => GA

bowei commented:
/lgtm
/approve

freehan commented:
/lgtm

bowei commented:
/lgtm

prameshj committed:
fixed golint error

exported func ToForwardingRuleComposite returns unexported type *gce.forwardingRuleComposite, which can be annoying to use

MrHohn reviewed (approved):
LGTM for the golint fix.
/lgtm
